[제이] 2단계 미션 제출합니다.

jjj0611 opened the pull request:
제이 2단계 로또 미션 제출합니다.

jjj0611 committed:
feat: 로또 개수를 입력 받는 count 객체 추가

jjj0611 committed:
refactoring. winningLotto 생성을 위한 dto를 만들고, inputview에서 dto를 반환하도록 변경

jjj0611 committed:
feat. dto로부터 domain객체를 생성하는 LottoService 생성

jjj0611 committed:
feat. 사용자에게 수동 구매를 위해 입력 받는 LottoTicketDto를 생성하고, LottoTicketDto로부터 LottoTicket을 생성하는 기능을 추가

jjj0611 committed:
feat. InputView와 GameParser의 역할을 분리하고 GameParser를 Service에서 관리하도록 변경했습니다.

jjj0611 committed:
refactoring. 커스텀 예외를 생성하여 예외 발생을 변경

jjj0611 committed:
refactoring. 불필요한 정보를 삭제

	- LottoFactory에서 LottoNumber의 변수를 가져오는 코드를 변경
	- LottoTickets test에서 Intstream을 사용하여 표현 간소화

jjj0611 committed:
refactoring. LottoTickets의 역할 변경 및 도메인 테스트 커버리지 달성

	- LottoTIckets에서 LottoResult를 생성하는로직을 LottoService에서 처리하도록 변경
	- InputView가 Domain객체를 받지 않고 원시 타입의 인자를 받도록 변경
	- 도메인 객체의 테스트 커버리지 달성

jeonghoon1107 commented on src/main/java/lotto/view/output/OutputView.java:
```diff
@@ -11,33 +11,34 @@
     }
 
     public static void printProfit(LottoResult lottoResult, LottoMoney lottoMoney) {
         double profit = lottoResult.getProfit(lottoMoney);
```
LottoResult를 생성할 때 구해도 좋을 것 같아요.

jeonghoon1107 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,75 @@
+        return new Count(lottoMoney.getLottoPurchaseCounts(), manualCount);
+    }
+
+    public LottoTickets createLottoTickets(List<String> lottoTicketInputs, Count count) {
```
자동로또, 수동로또 만드는 부분을 interface로 추출하는 연습을 해보면 어떨까요?
또한, LottoFactory를 이용해 이를 내부로 감출 수 있을 것 같아요!
LottoFactory가 LottoTickets을 반환하도록!

jjj0611 replied:
interface로 추출하고 LottoFactory 내부로 감추었습니다.

jeonghoon1107 commented on src/test/java/lotto/domain/count/CountTest.java:
```diff
@@ -0,0 +1,34 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class CountTest {
```
어떤 것을 테스트하는지 의도를 명확하게 하면 좋겠어요.

jjj0611 replied:
테스트명 수정 및 DisplayName 추가로 테스트의 의도를 표현했습니다.

jeonghoon1107 commented on src/main/java/lotto/parser/GameParser.java:
```diff
@@ -19,7 +19,7 @@ public int parseInputToInt(String input) {
         try {
             return Integer.parseInt(input);
         } catch (NumberFormatException ne) {
-            throw new IllegalArgumentException("숫자만 입력해주세요");
```
GameParser 클래스는 인스턴스 변수에 의존적이지 않은데 사용할 때마다 객체를 생성할 필요는 없을 것 같아요!

jjj0611 replied:
GameParser의 메서드를 static하게 변경하고, 인스턴스 생성을 하지 않고 사용했습니다.

jeonghoon1107 commented on src/test/java/lotto/domain/lotto/LottoNumberTest.java:
```diff
@@ -17,6 +18,15 @@
+    void equals() {
+        int expected = 4;
+        LottoNumber lottoNumber = new LottoNumber(expected);
+        assertThat(lottoNumber).isNotEqualTo(null);
```
원하는대로 테스트가 된 것인가요?

jjj0611 replied:
앗 이 부분은 equals를 오버라이드 하는 과정에서 test coverage를 100%로 맞추기 위해서 넣은 테스트였습니다.

eqauls에서 null과 비교하는 부분이 테스트를 안타서 저렇게 추가하게 되었습니다.
그런데 다시 작성하는 과정에서 equals를 굳이 오버라이드 할 필요가 없게 되어서 이 부분에 대한 테스트는 제거 하겠습니다 :)

jeonghoon1107 replied:
테스트 커버리지를 높이는 것은 중요하지만 의미 없는 테스트로 커버리지를 높이는 것은 지양해주세요!

jeonghoon1107 reviewed (changes requested):
안녕하세요 제이!
구현 잘 하셨네요 👍 
아쉬운 점이 있어서 피드백 남겼으니 확인해주세요 :)

jjj0611 committed:
refactoring. LottoNumber의 생성을 LottoFactory에서 LottoNumber로 변경

	- LottoFactory에서 LottoNumber를 불러서 생성하도록 변경
	- LottoNumber와 LottoTicket은 접근제어자를 변경해서 LottoFactory를 통해서만 접근하도록 변경
	- custom exception을 제거하고, 표준 exception으로 대체

jjj0611 committed:
refacotring. gameParser를 static 메서드로만 사용할 수 있도록 변경

jjj0611 committed:
refactoring. count를 test하는 메서드들의 의도를 명확하게 드러냄

jjj0611 committed:
refactoring. 수익률을 Outputview에서 계산하지 않고, 계산된 값을 넘겨주는 방식으로 변경

jjj0611 committed:
refacotring. LottoGenerator 인터페이스 생성 및 LottoFactory 변경

	- LottoGenerator 인터페이스를 생성
	- AutoGenerator와 ManualGenerator 생성
	- LottoFactory 내부에서 두 Generator를 사용하여 LottoTickets 생성

jjj0611 committed:
refactoring. LottoGenerator에 대한 test 추가

jeonghoon1107 commented on src/main/java/lotto/domain/lotto/generator/AutoLottoGenerator.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private LottoTicket generateLottoTicketOfRandom() {
+        Set<LottoNumber> lottoNumbers = random.ints(1, 46)
```
외부에서는 1, 46이라는 숫자를 몰라도 되게끔 Random을 사용하지 않고 LottoNumber에서 이미 존재하는 숫자들을 shuffle 후 6개를 꺼내와도 좋았을 것 같아요

jeonghoon1107 reviewed (approved):
안녕하세요 제이!
깔끔하네요 💯 💯 
미션 진행하느라 고생하셨습니다 :)
